lib/cn/kcompact.c

```diff
@@ -175,176 +79,212 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
     if (w->cw_prog_interval && w->cw_progress)
         tprog = jiffies;
 
-    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    err = bin_heap2_create(w->cw_kvset_cnt, kv_item_compare, &bh);
     if (ev(err))
         return err;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (!more || ev(err))
+    sources = malloc(w->cw_kvset_cnt * sizeof(*sources));
+    if (!sources) {
+        err = merr(ENOMEM);
         goto done;
+    }
 
-new_key:
+    for (uint i = 0; i < w->cw_kvset_cnt; i++)
+        sources[i] = kvset_iter_es_get(w->cw_inputv[i]);
 
-    if (atomic_read(w->cw_cancel_request)) {
-        err = merr(ev(ESHUTDOWN));
+    err = bin_heap2_prepare(bh, w->cw_kvset_cnt, sources);
+    if (ev(err))
         goto done;
-    }
 
-    if (tprog) {
-        u64 now = jiffies;
+    w->cw_stats.ms_srcs = w->cw_kvset_cnt;
+
+    more = bin_heap2_peek(bh, (void **)&curr);
+    while (more) {
+        uint idx;
+        struct kv_iterator *iter;
 
-        if (now - tprog > w->cw_prog_interval) {
-            tprog = now;
-            w->cw_progress(w);
+        if (atomic_read(w->cw_cancel_request)) {
+            err = merr(ESHUTDOWN);
+            goto done;
         }
-    }
 
-    emitted_val = false;
-    horizon = true;
-    emitted_seq = 0;
-    emitted_seq_pt = 0;
-
-    dbg_prev_seq = 0;
-    dbg_prev_src = 0;
-    dbg_nvals_this_key = 0;
-    dbg_dup = false;
-
-get_values:
-    vdata = NULL;
-
-    while (horizon &&
-           kvset_iter_next_vref(
-               w->cw_inputv[curr.src], &curr.vctx, &seq, &vtype, &vbidx, &vboff,
-               &vdata, &vlen, &complen))
-    {
-        bool should_emit = false;
-
-        /* Assertion logic:
-         *   if (dbg_nvals_this_key)
-         *       assert(dbg_prev_seq > seq);
-         */
-        if (HSE_UNLIKELY(dbg_nvals_this_key && dbg_prev_seq <= seq)) {
-            assert(0);
-            seqno_errcnt++;
+        iter = kvset_cursor_es_h2r(curr->src);
+        for (idx = 0; idx < w->cw_kvset_cnt; idx++) {
```

Comment:
This is not the cleanest. I could use an hse_util/map to get this to O(1). The only reason this exists in the first place is to index the vbmap. Nabeel, I think you added the vbmap fairly recently. Any ideas on how to clean this up? One idea that comes to mind is that the inputv could change to a struct containing both the kv_iterator and the vbmap.

Reply:
Won't this have the same perf issue that Nabeel remediated by using a binary search on vgmap?
Looks like a perf killer..

Reply:
Probably would. Hence the call for ideas!

Reply:
I'm fresh out!  I'll have to study the problem first, but maybe someone else has some insights.  Have you spent any time measuring the perf?

Reply:
How about updating kvset iterators with APIs to stash a value with the iterator that can be retrieved later?  Prior to the merge loop you could do this:
```
for (idx = 0; idx < w->cw_kvset_cnt; idx++)
    kvset_iter_set_context(w->cw_inputv[idx], (void*)idx);
```
And inside the merge loop you could get the index back with:
```
idx = (uint)kvset_iter_get_context(curr);
```

Reply:
That makes sense to me. Only issue is that this context literally exists for the purpose of assisting in testing. Very strange...

Reply:
For a good time use &idx !!!

Reply:
It's not just test/debug is it?   It's used later in the call to `kvset_builder_add_vref`

Reply:
Alternatively, since this binheap is created from an array of iterators, we could add a bin heap API to get the source number.

Reply:
Could you pre-lookup the vbmap and push that into the iter context to avoid the vbmap lookup?  Or vgmap lookup?  Both?    Probably I am just confused...

Reply:
> It's not just test/debug is it? It's used later in the call to `kvset_builder_add_vref`

Not kvi_context.

> What is the es_sort field in struct element source? Just wondering whether it has any relation to the kvset iterator ordering and if yes can we make use of it here?

Seems like `es_sort` is the index of the initial array of element sources. I think you are right Nabeel.

> Could you pre-lookup the vbmap and push that into the iter context to avoid the vbmap lookup? Or vgmap lookup? Both? Probably I am just confused...

Yes, I could, but this idx is also used in one of these debug variables that I really think should just be removed :).

> Alternatively, since this binheap is created from an array of iterators, we could add a bin heap API to get the source number.

This could come as part of the pop/peek APIs honestly.

The `struct kv_iterator::kvi_context` is really just being abused by tests that didn't know about `container_of()`. So I have changed the tests to not abuse it.

Given what Alex said above, we could just remove `kvi_context` in this PR if that is the route we wanted to go.

Reply:
Anything we can do to eliminate per-key non-constant time lookups is goodness, we're really hurting in this area right now..

Reply:
If you see something in the latest iteration, let me know